source/common/http/utility.cc

```diff
@@ -380,6 +382,94 @@ Utility::getLastAddressFromXFF(const Http::HeaderMap& request_headers, uint32_t
   }
 }
 
+void Utility::sanitizeConnectionHeader(spdlog::logger& logger, Http::HeaderMap& headers) {
+  // Remove any headers nominated by the Connection header. The TE header
+  // is sanitized and removed only if it's empty after removing unsupported values
+  // See https://github.com/envoyproxy/envoy/issues/8623
+  const auto& cv = Http::Headers::get().ConnectionValues;
+  const auto& connection_header_value = headers.Connection()->value();
+
+  StringUtil::CaseUnorderedSet headers_to_remove{};
+
+  // Split the connection header and evaluate each nominated header
+  for (const auto& token : absl::StrSplit(connection_header_value.getStringView(), ',')) {
+
+    const absl::string_view token_sv = StringUtil::trim(token);
+
+    // If the Connection token value is not a nominated header, ignore it here since
+    // the connection header is removed elsewhere when the H1 request is upgraded to H2
+    if (StringUtil::CaseInsensitiveCompare()(token_sv, cv.Close) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Http2Settings) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.KeepAlive) ||
+        StringUtil::CaseInsensitiveCompare()(token_sv, cv.Upgrade)) {
+      continue;
+    }
+
+    // Build the LowerCaseString for header lookup
+    const std::string header_to_remove = std::string(token_sv.data(), token_sv.size());
+    const LowerCaseString lcs_header_to_remove(header_to_remove);
+
+    // By default we will remove any nominated headers
+    bool keep_header = false;
+    StringUtil::CaseUnorderedSet tokens_to_remove{};
+
+    // Determine whether the nominated header contains unsupported values
+    HeaderEntry* nominated_header = headers.get(lcs_header_to_remove);
+
+    if (nominated_header) {
```

Comment:
Let me take a look and see how I can avoid this.

Reply:
Sorry for not making this more clear but I don't think we can avoid the m x n x p factor, I think we instead need to limit m, n, and p.
I think we already added configurable limits on the number of headers.  I'd be inclined to also limit the number of connection and/or TE tokens (configurable).  I've never seen more than 3 connection tokens in valid HTTP.  I think if we cap it at something an order of magnitude larger (10?  30?) we won't drop legit traffic but we will remove that attack vector.  I suspect we'd have to make it configurable just in case someone has valid traffic with a ton of headers. 
@mattklein123 WDYT?